### ISSUE TITLE
Formalize or remove OP2Utility.h

### ISSUE THREAD

**Issue by Brett208** [OutpostUniverse/OP2Utility#108]
Currently, there is an unmaintained library-wide header file called OP2Utility.h. It is used in about 3 places in OP2Archive to bring in OP2Utility.

Do we want to support a library wide header that brings in all expected to be public facing elements of the library, or is it better for users to specify individual headers in the library?

At first I was against maintaining this, but now I'm not so sure. It can be nice to see the highpoint headers all listed in one place when first being introduced into the library. 

I guess the contents of the header file would need to be:
* Archives: 
  * ArchiveFile.h
  * ClmFile.h
  * VolFile.h
  * ~~CompressionTypes.h~~ (Already brought in by VolFile.h) 
* Maps
  * MapData.h
  * MapReader.h
  * MapWriter.h
* Streams 
  * FileStreamReader.h
  * FileStreamWriter.h
  * FileSliceReader.h
  * MemoryStreamReader.h
  * MemoryStreamWriter.h
* Utility
  * ResourceManager.h
  * XFile.h
  * StreamHelper.h

Thoughts?

-Brett

**Comment by DanRStevens** [comment]
I think this nicely illustrates the difference between `src/` and `include/`. The internal headers, used within the library will be mixed in with `src/`. These need to be maintained while working on the library so the library can be compiled. The import headers, which contain the public stuff for users of the library will end up in `include/`. It needs to be maintained with each release, so users of the library can access the functionality and allow client code to be compiled. The library itself might never reference anything in `include/`.

As for module breakdown, I think you did a pretty good job there. In particular, CompressionTypes.h isn't really usable standalone. You'd only ever use it through VolFile.h. Whether you want to support a high level library include, such as "op2utility", or module specific includes, such as "Archive", or individual components such as "VolFile" is another matter. I'm a little uncertain about my recommendation here. I'm tempted to suggest module level for now, such as "Map", "Archive", "Stream", "Utility".

Looking at this though, it seems like "Stream" might want to support more granularity. There is quite a difference in the stream types. You won't likely need all stream types that can be offered, and the main thing tying them together is a common interface. Though I think this might simply be an indication that "Stream" could be it's own independent library. I don't much want to deal with that complication right now though. I'd rather just leave it in OP2Utility, and perhaps access it through a module level include.

I think maybe have module level includes, and a full library include, that includes each of the modules. We probably won't know what we should do until we've tried something.

**Comment by Brett208** [comment]
We seem in agreement to push a /include directory that contains external facing header files. I like the idea of a few major modules and then a master header that has everything.

I'm thinking Map.h is too generic. It could mean an actual map construct. OP2Map.h sounds better. I also like the idea of OP2Archive.h as it would specify that the archives are meant for Outpost 2, especially if someone has some other archive code going on? (maybe that doesn't really exist).

I'd probably prefer to avoid having both a Utility.h and OP2Utility.h header. Wher Utility.h had XFile and StringHelper, etc, while OP2Utility was the master header.

The Stream code isn't OP2 specifc, so I am happy with Stream.h, but this seems really generic and could conflict with other libraries?

@DanRStevens, do you think it is important to pick external facing header filenames that are distince from possibilities in other libraries, or does it not matter?

If I can fix my line ending woes in this project, I'll take a crack at this.

Thanks,
Brett

**Comment by Brett208** [comment]
I've thought about this more and am disagreeing stronger now. I think they should be

* OP2Utility - Includes everything
* OP2Map
* OP2Archive
* Stream
* ??? (I'm not sure Utility.h works here since we have OP2Utility.)

Typing OP2Map or OP2Archive is minimal more typing with more clarity on the source. If someone is using a C++ map structure for some reason and map.h from us, it seems overly confusing as they would have to decipher which was which. OP2Map would instantly tell them it was an Outpost related Map include and not a map data structure. 

Same with Archive. Is this a generic archive that uses a zip format? I think it instantly tells me that someone is including Outpost 2 specific archive files by naming it OP2Archive. 

I feel like calling the entire project utility would be a bit vague. Maybe I'm getting too used to auto-fill on Visual Studio helping out though.

-Brett

**Comment by DanRStevens** [comment]
I don't think there would be any confusion if the includes were:
```
#include "OP2Utility/Map.h"
#include "OP2Utility/Archive.h"
```

What I want to avoid is useless redundancy, which might appear as:
```
#include "OP2Utility/OP2Map.h"

OP2Utility::OP2Map::MapData map;
```

Is this concern perhaps more about a lack `namespace` being used in the headers?

**Comment by Brett208** [comment]
Sorry, I skipped some of my logic in the last post. I'm still figuring out the best practices for library includes.

In Visual Studio, I can specify additional include directories for a project. In the case of OP2Archive, I would like to add the following directory as an include directory: `./OP2Utility/include`. I assume something similar is possible with CMake and Makefiles.

Once this is done, then inside of OP2Archive, I would just type `#include<OP2Archive.h>`. This way external library include files would not need any sort of directories added.

I think this is the right answer. If so, then I want the names of Archive and Map to include OP2.

If we want to make the user include the full relative directory, then I think it would make sense to drop the OP2 part of map and the others because it is already covered by the directory. But I think the average user of the library will just add the include directory to their build process and not prepend the relative directory to header includes. I also do not think we have control over whether they just include the include directory or not, so the more conservative approach would be to label the file itself.

How does this sound?

Thanks,
Brett

**Comment by DanRStevens** [comment]
What you say makes sense.

Still, my intuition is tingling on this one. I suspect more experienced C++ library distributors would object to something here. I think BlackBox objected to a similar idea I had years ago.

I think what bothers me is the number of header files. If there is a single header file, it should go in `$prefix/include/`. If there are multiple header files, they go in `$prefix/include/ProjectName/`.

On Linux, when you install stuff, they don't go in a project specific folder. They get mixed together as part of the system install. The `$prefix` is the install location, where all libraries and applications are installed under. A common prefix might be `/usr`, hence executable files appear in `/usr/bin/`, while libraries might appear in `/usr/lib/`, and header files in `/usr/include/`. Binaries from all sorts of packages are all mixed together in `/usr/bin/`. Same for libraries and headers in the other folders. There is no project specific folder containing all files.

Now I'll certainly admit the Linux way of doing things strikes me as a bit messy, but it is what it is, and it works. It also has a few advantages. You generally don't need to update the `$PATH` environment variable when you install something new, since all (most) binaries end up in the same `/usr/bin/` folder. Similarly, when compiling code, you don't usually need to set a specific include folder, since all (most) headers appear in `/usr/include/`.

As a consequence to this design, you probably want to avoid polluting the `include/` folder with a lot of different header files. Most packages have a single header file that is installed to that folder. If you want to install more than one file, you'd probably put them in a subfolder. SDL does this, and so you have a `/usr/include/SDL/` folder, which contains a bunch of header files. Either includes are of the form `#include <SDL/SDL.h>`, or an additional include folder is added to the compiler options `-I /usr/include/SDL` and the source uses `#include <SDL.h>`. If a different SDL header file is accessed directly, perhaps one of the other SDL modules, such as SDL-mixer, it would either have a folder prefix in the `#include` line, or it would have an additional include folder at the command line for the `/usr/include/SDL/` folder. Hence the built in disambiguation.

I think that's the structure we should model after. If we have multiple header files, we should probably have a project name subfolder within the include folder to hold them all: `include/OP2Utility/`

**Comment by Brett208** [comment]
Outside of any revelation from leeor, I would like to close this issue. The current repo includes an updated OP2Utility.h, which solves this issue. We can use this header in OP2Archive and OP2MapImager. If we want to further subdivide the external header into smaller ones in the future, we can revisit by opening an issue for it. I don't see the current discussion leading anywhere constructive.

-Brett

**Comment by DanRStevens** [comment]
One final matter that just occurred to me while reviewing the code move to `src/`. The current `include/OP2Utility.h` file references the headers from `src/`. For a binary package distribution, there will be no `src/` folder.

Perhaps the `include/OP2Utility.h` file should be auto generated by concatenating all the header files from the project. That way you have a self contained header, which doesn't depend on anything else. Plus, the construction should be a formulaic include all, which is perhaps better done by a machine. One important point, the concatenation would need to be done in include order, so as to ensure definitions are available before they are used.

**Comment by DanRStevens** [comment]
I think you're right in that the immediate topic of this issue has been addressed. We probably want a public facing import header. We now have one. Issues as to it's structure or maintenance are another matter.

We should probably close this.
